<a href="https://colab.research.google.com/github/OweT1/personal/blob/main/NUS_FinTech.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Importing of Packages

In [ ]:
#!pip install pandas_market_calendars
#import pandas_market_calendars as mcal

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import datetime as dt
from dateutil.relativedelta import relativedelta

# Reading in of Data (Using Google Colab & Google Drive)

In [ ]:
from google.colab import drive
drive.mount("/content/drive")

df = pd.read_csv("/content/drive/MyDrive/Colab Datasets/NUS FinTech/companies_pricing_data.csv")

df.head()

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


DATE TICKER                  SECTOR  PRICEOPEN  PRICEHIGH  PRICELOW  \
0  4/1/21   AAPL  Information Technology     133.52   133.6116   126.760   
1  5/1/21   AAPL  Information Technology     128.89   131.7400   128.430   
2  6/1/21   AAPL  Information Technology     127.72   131.0499   126.382   
3  7/1/21   AAPL  Information Technology     128.36   131.6300   127.860   
4  8/1/21   AAPL  Information Technology     132.43   132.6300   130.230   

   PRICECLOSE     VOLUME  
0      129.41  143301900  
1      131.01   97664900  
2      126.60  155088000  
3      130.92  109578200  
4      132.05  105158200

In [ ]:
df.dtypes

DATE           object
TICKER         object
SECTOR         object
PRICEOPEN     float64
PRICEHIGH     float64
PRICELOW      float64
PRICECLOSE    float64
VOLUME          int64
dtype: object

In [ ]:
df["TICKER"].unique()

array(['AAPL', 'AMZN', 'APD', 'BA', 'BAC', 'BRK.A', 'CBRE', 'CLST',
       'CMCS.A', 'COP', 'COST', 'CVX', 'D', 'DIS', 'DUK', 'ECL', 'EOG',
       'EXC', 'FCX', 'GOOGL', 'HD', 'HON', 'JLL', 'JNJ', 'JPM', 'KO',
       'LIN', 'LLY', 'MA', 'MCD', 'META', 'MSFT', 'NEE', 'NFLX', 'NKE',
       'NVDA', 'OPEN', 'PEP', 'PFE', 'PG', 'PXD', 'RTX', 'SHW', 'SO',
       'TMO', 'TSLA', 'UNH', 'UNP', 'UPS', 'V', 'WE', 'WFC', 'WMT', 'XOM',
       'ZG'], dtype=object)

In [ ]:
print(df["DATE"].iloc[0])
print(df["DATE"].iloc[-1])

4/1/21
31/12/21


In [ ]:
print(df.shape)

(13664, 8)


# Q1
Generate a new column "PRICEMID", which is taken as the simple average of the "PRICEHIGH" and "PRICELOW" columns. (12.5 points)

## Approach:
Simply take the average of the 2 columns.

In [ ]:
df["PRICEMID"] = (df["PRICEHIGH"] + df["PRICELOW"])/2

df.head(10)

DATE TICKER                  SECTOR  PRICEOPEN  PRICEHIGH  PRICELOW  \
0   4/1/21   AAPL  Information Technology     133.52   133.6116   126.760   
1   5/1/21   AAPL  Information Technology     128.89   131.7400   128.430   
2   6/1/21   AAPL  Information Technology     127.72   131.0499   126.382   
3   7/1/21   AAPL  Information Technology     128.36   131.6300   127.860   
4   8/1/21   AAPL  Information Technology     132.43   132.6300   130.230   
5  11/1/21   AAPL  Information Technology     129.19   130.1700   128.500   
6  12/1/21   AAPL  Information Technology     128.50   129.6900   126.860   
7  13/1/21   AAPL  Information Technology     128.76   131.4500   128.490   
8  14/1/21   AAPL  Information Technology     130.80   131.0000   128.760   
9  15/1/21   AAPL  Information Technology     128.78   130.2242   127.000   

   PRICECLOSE     VOLUME   PRICEMID  
0      129.41  143301900  130.18580  
1      131.01   97664900  130.08500  
2      126.60  155088000  128.71595  
3      130.92  109578200  129.74500  
4      132.05  105158200  131.43000  
5      128.98  100620900  129.33500  
6      128.80   91951150  128.27500  
7      130.89   88636830  129.97000  
8      128.91   90221760  129.88000  
9      127.14  111598500  128.61210

# Q2
Generate a new column "SMA_5", which calculates the 5-trading-day Simple Moving Average of each companies' stock (for a stock ticker, at the end of each trading day, take the average of the past 4 trading days' "PRICECLOSE" along with that day's "PRICECLOSE"). Note that it is expected that the first 4 trading days of each ticker to not have a defined "SMA_5". (12.5 points)

## Approach:
Use the `.rolling()` function to calculate the 5-trading-day moving average. We note that since this rolling average is for each stock ticker, we will need to "partition" by the `TICKER` column.

In [ ]:
df["SMA_5"] = df.groupby("TICKER")["PRICECLOSE"].transform(lambda x: x.rolling(5).mean())

print(df.head())

# Checking to see that SMA_5 are NaN for the first 4 dates of each stock
print(df[df["SMA_5"].isna()].head(10))

     DATE TICKER                  SECTOR  PRICEOPEN  PRICEHIGH  PRICELOW  \
0  4/1/21   AAPL  Information Technology     133.52   133.6116   126.760   
1  5/1/21   AAPL  Information Technology     128.89   131.7400   128.430   
2  6/1/21   AAPL  Information Technology     127.72   131.0499   126.382   
3  7/1/21   AAPL  Information Technology     128.36   131.6300   127.860   
4  8/1/21   AAPL  Information Technology     132.43   132.6300   130.230   

   PRICECLOSE     VOLUME   PRICEMID    SMA_5  
0      129.41  143301900  130.18580      NaN  
1      131.01   97664900  130.08500      NaN  
2      126.60  155088000  128.71595      NaN  
3      130.92  109578200  129.74500      NaN  
4      132.05  105158200  131.43000  129.998  
       DATE TICKER                  SECTOR  PRICEOPEN  PRICEHIGH  PRICELOW  \
0    4/1/21   AAPL  Information Technology   133.5200  133.61160   126.760   
1    5/1/21   AAPL  Information Technology   128.8900  131.74000   128.430   
2    6/1/21   AAPL  Informa

# Q3
Generate the "DAILY_RETURN", which calculates the 1-day return for each stock (in percentage) at each trading day. The 1-day return of today is calculated based on the "PRICECLOSE" of the most recent trading day against the "PRICECLOSE" of its previous trading day. Express values as a percentage, e.g. a 21.0% return should have the value 21.0 (12.5 points)

## Approach:
We can simply "lag" the `PRICECLOSE` column by using the `.shift()` function. We still need to take note of the partitioning between the various stocks.

In [ ]:
df["PREV_CLOSE"] = df.groupby("TICKER")["PRICECLOSE"].transform(lambda x: x.shift(1))
df["DAILY_RETURN"] = round((df["PRICECLOSE"] - df["PREV_CLOSE"])/df["PREV_CLOSE"] * 100, 1)
df = df.drop(["PREV_CLOSE"], axis = 1)

print(df.head())

# Checking that all the DAILY_RETURN are NaN for all the first dates of each stock
print(df[df["DAILY_RETURN"].isna()].head())

     DATE TICKER                  SECTOR  PRICEOPEN  PRICEHIGH  PRICELOW  \
0  4/1/21   AAPL  Information Technology     133.52   133.6116   126.760   
1  5/1/21   AAPL  Information Technology     128.89   131.7400   128.430   
2  6/1/21   AAPL  Information Technology     127.72   131.0499   126.382   
3  7/1/21   AAPL  Information Technology     128.36   131.6300   127.860   
4  8/1/21   AAPL  Information Technology     132.43   132.6300   130.230   

   PRICECLOSE     VOLUME   PRICEMID    SMA_5  DAILY_RETURN  
0      129.41  143301900  130.18580      NaN           NaN  
1      131.01   97664900  130.08500      NaN           1.2  
2      126.60  155088000  128.71595      NaN          -3.4  
3      130.92  109578200  129.74500      NaN           3.4  
4      132.05  105158200  131.43000  129.998           0.9  
        DATE TICKER                  SECTOR  PRICEOPEN  PRICEHIGH  PRICELOW  \
0     4/1/21   AAPL  Information Technology     133.52   133.6116  126.7600   
252   4/1/21   AMZN

# Q4
Generate the "MONTHLY_RETURN", which calculates the 30-day return for each stock (in percentage) at each trading day. Note that this is 30 calendar days, where non-trading days take on the values of their last trading day. Similar to Q3, reference prices from the "PRICECLOSE" column. Express values as a percentage, e.g. a 21.0% return should have the value 21.0 (12.5 points)


## Approach:
We will first find the date that is 30 days before the date in each row. We will then check to see if that date is in the data (>= 4/1/21) and is a trading day. With this date, we can then find the index of which this date is represented in the data for each stock, and thereafter extract the `PRICECLOSE` for the date.

In [ ]:
df["NEW_DATE"] = pd.to_datetime(df["DATE"], format = "%d/%m/%y") # convert to datetime format
#df["NEW_DATE"] = [dt.datetime(year = int("20" + date.split("/")[2]), month = int(date.split("/")[1]), day = int(date.split("/")[0])) for date in df["DATE"]]
df["DATE_30DAYBEFORE"] = [date + relativedelta(days = -30) for date in df["NEW_DATE"]] # get the "30th day before" for the date in each row

print(df.head())

     DATE TICKER                  SECTOR  PRICEOPEN  PRICEHIGH  PRICELOW  \
0  4/1/21   AAPL  Information Technology     133.52   133.6116   126.760   
1  5/1/21   AAPL  Information Technology     128.89   131.7400   128.430   
2  6/1/21   AAPL  Information Technology     127.72   131.0499   126.382   
3  7/1/21   AAPL  Information Technology     128.36   131.6300   127.860   
4  8/1/21   AAPL  Information Technology     132.43   132.6300   130.230   

   PRICECLOSE     VOLUME   PRICEMID    SMA_5  DAILY_RETURN   NEW_DATE  \
0      129.41  143301900  130.18580      NaN           NaN 2021-01-04   
1      131.01   97664900  130.08500      NaN           1.2 2021-01-05   
2      126.60  155088000  128.71595      NaN          -3.4 2021-01-06   
3      130.92  109578200  129.74500      NaN           3.4 2021-01-07   
4      132.05  105158200  131.43000  129.998           0.9 2021-01-08   

  DATE_30DAYBEFORE  
0       2020-12-05  
1       2020-12-06  
2       2020-12-07  
3       2020-12-08  

In [ ]:
print(df["DATE_30DAYBEFORE"])

0       2020-12-05
1       2020-12-06
2       2020-12-07
3       2020-12-08
4       2020-12-09
           ...    
13659   2021-11-27
13660   2021-11-28
13661   2021-11-29
13662   2021-11-30
13663   2021-12-01
Name: DATE_30DAYBEFORE, Length: 13664, dtype: datetime64[ns]


In [ ]:
start_date = min(df["NEW_DATE"]) # 2021-01-04
end_date = max(df["NEW_DATE"]) # 2021-12-31

# nyse = mcal.get_calendar('NYSE')
# trading_days = nyse.valid_days(start_date=start_date, end_date=end_date)

lost_dates = sorted(list(filter(lambda x: x >= df["NEW_DATE"][0], set(df["DATE_30DAYBEFORE"]).difference(set(df["NEW_DATE"]))))) # dates that are not in the date range (ie earlier than 4/1/21 or not a trading day)

# We will create a mapping to map the "30th day before" to the latest trading day
lost_dates_mapping = {}

for lost_date in lost_dates:
  diff = -1
  while lost_date not in lost_dates_mapping:
    new_date = lost_date + relativedelta(days = diff)
    if new_date in list(df["NEW_DATE"]):
      lost_dates_mapping[lost_date] = new_date
    else:
      diff -= 1

print(lost_dates_mapping)

{Timestamp('2021-01-09 00:00:00'): Timestamp('2021-01-08 00:00:00'), Timestamp('2021-01-10 00:00:00'): Timestamp('2021-01-08 00:00:00'), Timestamp('2021-01-17 00:00:00'): Timestamp('2021-01-15 00:00:00'), Timestamp('2021-01-18 00:00:00'): Timestamp('2021-01-15 00:00:00'), Timestamp('2021-01-23 00:00:00'): Timestamp('2021-01-22 00:00:00'), Timestamp('2021-01-24 00:00:00'): Timestamp('2021-01-22 00:00:00'), Timestamp('2021-01-30 00:00:00'): Timestamp('2021-01-29 00:00:00'), Timestamp('2021-01-31 00:00:00'): Timestamp('2021-01-29 00:00:00'), Timestamp('2021-02-06 00:00:00'): Timestamp('2021-02-05 00:00:00'), Timestamp('2021-02-07 00:00:00'): Timestamp('2021-02-05 00:00:00'), Timestamp('2021-02-13 00:00:00'): Timestamp('2021-02-12 00:00:00'), Timestamp('2021-02-14 00:00:00'): Timestamp('2021-02-12 00:00:00'), Timestamp('2021-02-15 00:00:00'): Timestamp('2021-02-12 00:00:00'), Timestamp('2021-02-20 00:00:00'): Timestamp('2021-02-19 00:00:00'), Timestamp('2021-02-21 00:00:00'): Timestamp('20

In [ ]:
# We will map the new dates to a new column
df["DATE_30DAYBEFORE_NEW"] = pd.Series(map(lambda x: lost_dates_mapping[x] if x in lost_dates_mapping else x, df["DATE_30DAYBEFORE"]))

df[df["DATE"] == "23/2/21"].head() # 2021-01-24 is a Sunday, the new date, 2021-01-22, is a Friday

DATE TICKER                  SECTOR  PRICEOPEN  PRICEHIGH  PRICELOW  \
34    23/2/21   AAPL  Information Technology   123.7600   126.7100    118.39   
286   23/2/21   AMZN  Consumer Discretionary   156.3515   160.2365    154.68   
538   23/2/21    APD               Materials   265.6500   267.4900    259.59   
790   23/2/21     BA             Industrials   214.1500   214.7500    206.60   
1042  23/2/21    BAC              Financials    35.4200    35.8900     35.14   

      PRICECLOSE     VOLUME   PRICEMID     SMA_5  DAILY_RETURN   NEW_DATE  \
34       125.860  158273000  122.55000  128.4560          -0.1 2021-02-23   
286      159.725   93544440  157.45825  162.6201           0.4 2021-02-23   
538      263.330    1335653  263.54000  263.0700          -0.8 2021-02-23   
790      212.120   15996700  210.67500  213.2940          -0.4 2021-02-23   
1042      35.520   67440700   35.51500   34.7620           1.1 2021-02-23   

     DATE_30DAYBEFORE DATE_30DAYBEFORE_NEW  
34         2021-01-24           2021-01-22  
286        2021-01-24           2021-01-22  
538        2021-01-24           2021-01-22  
790        2021-01-24           2021-01-22  
1042       2021-01-24           2021-01-22

In [ ]:
#df["DATE_30DAYBEFORE_NEW_IND"] = df.groupby("TICKER")["DATE_30DAYBEFORE_NEW"].transform(lambda x: df[df["NEW_DATE"] == x].index.tolist())
# df_copy = df
# df["DATE_30DAYBEFORE_NEW_IND"] = [df[(row["TICKER"] == new_df["TICKER"] & row["DATE_30DAYBEFORE_NEW"] == new_df["NEW_DATE"]) for row in df.rows].index]

In [ ]:
# We will get the index of which the 30th day before for each stock exists in the data. If it does not exist, we will replace it with NaN
def get_index(date, ticker):
    return df[(df['NEW_DATE'] == date) & (df['TICKER'] == ticker)].index

df["DATE_30DAYBEFORE_NEW_IND"] = df.apply(lambda row: get_index(row['DATE_30DAYBEFORE_NEW'], row['TICKER']).to_list(), axis=1)
df["DATE_30DAYBEFORE_NEW_IND"] = [int(ind[0]) if ind != [] else np.nan for ind in df["DATE_30DAYBEFORE_NEW_IND"]]
df[df["DATE"] == "23/2/21"].head()

DATE TICKER                  SECTOR  PRICEOPEN  PRICEHIGH  PRICELOW  \
34    23/2/21   AAPL  Information Technology   123.7600   126.7100    118.39   
286   23/2/21   AMZN  Consumer Discretionary   156.3515   160.2365    154.68   
538   23/2/21    APD               Materials   265.6500   267.4900    259.59   
790   23/2/21     BA             Industrials   214.1500   214.7500    206.60   
1042  23/2/21    BAC              Financials    35.4200    35.8900     35.14   

      PRICECLOSE     VOLUME   PRICEMID     SMA_5  DAILY_RETURN   NEW_DATE  \
34       125.860  158273000  122.55000  128.4560          -0.1 2021-02-23   
286      159.725   93544440  157.45825  162.6201           0.4 2021-02-23   
538      263.330    1335653  263.54000  263.0700          -0.8 2021-02-23   
790      212.120   15996700  210.67500  213.2940          -0.4 2021-02-23   
1042      35.520   67440700   35.51500   34.7620           1.1 2021-02-23   

     DATE_30DAYBEFORE DATE_30DAYBEFORE_NEW  DATE_30DAYBEFORE_NEW_IND  
34         2021-01-24           2021-01-22                      13.0  
286        2021-01-24           2021-01-22                     265.0  
538        2021-01-24           2021-01-22                     517.0  
790        2021-01-24           2021-01-22                     769.0  
1042       2021-01-24           2021-01-22                    1021.0

In [ ]:
# We will use the index to extract the `PRICECLOSE` for the respective dates
df["PRICE_30DAYBEFORE"] = [df["PRICECLOSE"][ind] if not np.isnan(ind) else np.nan for ind in df["DATE_30DAYBEFORE_NEW_IND"]]

# Checking of `PRICECLOSE`
print(df[df["DATE"] == "23/2/21"].head())
print(df[df["DATE"] == "22/1/21"].head())

         DATE TICKER                  SECTOR  PRICEOPEN  PRICEHIGH  PRICELOW  \
34    23/2/21   AAPL  Information Technology   123.7600   126.7100    118.39   
286   23/2/21   AMZN  Consumer Discretionary   156.3515   160.2365    154.68   
538   23/2/21    APD               Materials   265.6500   267.4900    259.59   
790   23/2/21     BA             Industrials   214.1500   214.7500    206.60   
1042  23/2/21    BAC              Financials    35.4200    35.8900     35.14   

      PRICECLOSE     VOLUME   PRICEMID     SMA_5  DAILY_RETURN   NEW_DATE  \
34       125.860  158273000  122.55000  128.4560          -0.1 2021-02-23   
286      159.725   93544440  157.45825  162.6201           0.4 2021-02-23   
538      263.330    1335653  263.54000  263.0700          -0.8 2021-02-23   
790      212.120   15996700  210.67500  213.2940          -0.4 2021-02-23   
1042      35.520   67440700   35.51500   34.7620           1.1 2021-02-23   

     DATE_30DAYBEFORE DATE_30DAYBEFORE_NEW  DATE_30DAYBE

In [ ]:
df["MONTHLY_RETURN"] = round((df["PRICECLOSE"] - df["PRICE_30DAYBEFORE"])/df["PRICE_30DAYBEFORE"] * 100, 1)

print(df[df["DATE"] == "23/2/21"].head())

         DATE TICKER                  SECTOR  PRICEOPEN  PRICEHIGH  PRICELOW  \
34    23/2/21   AAPL  Information Technology   123.7600   126.7100    118.39   
286   23/2/21   AMZN  Consumer Discretionary   156.3515   160.2365    154.68   
538   23/2/21    APD               Materials   265.6500   267.4900    259.59   
790   23/2/21     BA             Industrials   214.1500   214.7500    206.60   
1042  23/2/21    BAC              Financials    35.4200    35.8900     35.14   

      PRICECLOSE     VOLUME   PRICEMID     SMA_5  DAILY_RETURN   NEW_DATE  \
34       125.860  158273000  122.55000  128.4560          -0.1 2021-02-23   
286      159.725   93544440  157.45825  162.6201           0.4 2021-02-23   
538      263.330    1335653  263.54000  263.0700          -0.8 2021-02-23   
790      212.120   15996700  210.67500  213.2940          -0.4 2021-02-23   
1042      35.520   67440700   35.51500   34.7620           1.1 2021-02-23   

     DATE_30DAYBEFORE DATE_30DAYBEFORE_NEW  DATE_30DAYBE

In [ ]:
# Dropping of columns introduced earlier
new_columns = ["NEW_DATE", "DATE_30DAYBEFORE", "DATE_30DAYBEFORE_NEW", "DATE_30DAYBEFORE_NEW_IND", "PRICE_30DAYBEFORE"]
df = df.drop(new_columns, axis = 1)

print(df.head())
print(df.shape)

     DATE TICKER                  SECTOR  PRICEOPEN  PRICEHIGH  PRICELOW  \
0  4/1/21   AAPL  Information Technology     133.52   133.6116   126.760   
1  5/1/21   AAPL  Information Technology     128.89   131.7400   128.430   
2  6/1/21   AAPL  Information Technology     127.72   131.0499   126.382   
3  7/1/21   AAPL  Information Technology     128.36   131.6300   127.860   
4  8/1/21   AAPL  Information Technology     132.43   132.6300   130.230   

   PRICECLOSE     VOLUME   PRICEMID    SMA_5  DAILY_RETURN  MONTHLY_RETURN  
0      129.41  143301900  130.18580      NaN           NaN             NaN  
1      131.01   97664900  130.08500      NaN           1.2             NaN  
2      126.60  155088000  128.71595      NaN          -3.4             NaN  
3      130.92  109578200  129.74500      NaN           3.4             NaN  
4      132.05  105158200  131.43000  129.998           0.9             NaN  
(13664, 12)


# Q5
Extracting of new Dataframe into a `.csv` file

In [ ]:
df.to_csv("e0958335.csv", index = False)

# Q6
What is the standard deviation of the PRICECLOSE column across all trading days and tickers? (rounded to the nearest integer)

In [ ]:
print(df["PRICECLOSE"].describe()) # Unbiased
print(np.std(df["PRICECLOSE"])) # Biased

count     13664.000000
mean       7803.474366
std       55380.173001
min           7.530000
25%          74.177500
50%         161.391250
75%         259.572500
max      454550.000000
Name: PRICECLOSE, dtype: float64
55378.146464359976


Answer: 55378 (Biased estimate of standard deviation), 55380 (Unbiased estimate of standard deviation)

# Q7
What is the minimum value of the PRICELOW column across all trading days and tickers? (rounded to the nearest 2 d.p.)

In [ ]:
print(df["PRICELOW"].describe())
print(min(df["PRICELOW"]))

count     13664.000000
mean       7750.967318
std       55016.280217
min           7.200000
25%          73.210000
50%         159.830000
75%         256.690000
max      452545.140000
Name: PRICELOW, dtype: float64
7.2


Answer: 7.20

# Q8
What is the TICKER of the stock with the lowest DAILY_RETURN in the "Financials" sector as of 30 June 2021 (i.e. 1 Jan 2021 - 30 June 2021 inclusive)?


In [ ]:
df_q8 = df[(df["DATE"] > "1/1/21") & (df["DATE"] <= "30/6/2021") & (df["SECTOR"] == "Financials")]
print(df_q8[df_q8["DAILY_RETURN"] == df_q8["DAILY_RETURN"].min()])

          DATE TICKER      SECTOR  PRICEOPEN  PRICEHIGH  PRICELOW  PRICECLOSE  \
12665  15/1/21    WFC  Financials      33.08       33.5     31.82       32.04   

         VOLUME  PRICEMID   SMA_5  DAILY_RETURN  MONTHLY_RETURN  
12665  91993100     32.66  33.554          -7.8             NaN  


Answer: WFC

# Q9
What is the TICKER of the stock with the highest MONTHLY_RETURN in the "Information Technology" sector across all trading days?

In [ ]:
df_q9 = df[(df["SECTOR"] == "Information Technology")]
print(df_q9[df_q9["MONTHLY_RETURN"] == df_q9["MONTHLY_RETURN"].max()])

          DATE TICKER                  SECTOR  PRICEOPEN  PRICEHIGH  PRICELOW  \
8847  19/11/21   NVDA  Information Technology     322.43     330.88    319.05   

      PRICECLOSE    VOLUME  PRICEMID    SMA_5  DAILY_RETURN  MONTHLY_RETURN  
8847      329.85  53386670   324.965  308.298           4.1            49.2  


Answer: NVDA

# Q10
Which is the best-performing sector (sector with the highest average MONTHLY_RETURN across all its stocks) as of 31 December 2021 (i.e. 1 Jan 2021 - 31 December 2021 inclusive)?

In [ ]:
df_q10 = df.groupby("SECTOR")["MONTHLY_RETURN"].mean()

print(df_q10)

SECTOR
Communication Services    1.437835
Consumer Discretionary    2.285714
Consumer Staples          1.695844
Energy                    4.033593
Financials                2.869520
Health Care               2.721645
Industrials               1.350823
Information Technology    3.204416
Materials                 2.120693
Real Estate              -0.319048
Utilities                 1.297143
Name: MONTHLY_RETURN, dtype: float64


Answer: Energy